btree/btree.c

```diff
@@ -2939,145 +2953,128 @@ static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
+static void ff_node_update_credit(const struct nd *node,
+				   struct m0_be_tx_credit *cred)
 {
-	accum->tc_cb_nr += BTREE_CALLBACK_CREDIT;
+	int         shift     = ff_shift(node);
+	m0_bcount_t node_size = 1ULL << shift;
+
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, node_size));
 }
 
 /**
- * This function will calculate credits required to allocate node and it will
- * add those credits to @accum.
+ * This function will calculate credits required to free the node and it will
+ *  add those credits to @cred.
  */
-static void btree_node_alloc_credit(const struct m0_btree  *tree,
-				    struct m0_be_tx_credit *accum)
+static void ff_node_free_credit(const struct nd *node,
+				struct m0_be_tx_credit *cred)
 {
-	m0_bcount_t             node_size;
-	int                     shift;
+	struct ff_head *h           = ff_data(node);
+	int             shift       = h->ff_shift;
+	int             node_size   = 1ULL << shift;
+	int             header_size = sizeof(*h);
 
-	shift     = node_shift(tree->t_desc->t_root);
-	node_size =  1ULL << shift;
+	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED,
+			       node_size, shift, cred);
 
-	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
-			       node_size, shift, accum);
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, header_size));
 }
 
 /**
- * This function will calculate credits required to update node and it will add
+ * This function will calculate credits required to split node and it will add
  * those credits to @accum.
  */
-static void btree_node_update_credit(const struct m0_btree  *tree,
-				     struct m0_be_tx_credit *accum)
+static void ff_node_split_credit(const struct nd *node,
+				 struct m0_be_tx_credit *cred)
 {
- 	m0_bcount_t             node_size;
-	int                     shift;
+	struct m0_be_tx_credit cred_update_node = {};
 
-	shift     = node_shift(tree->t_desc->t_root);
-	node_size =  1ULL << shift;
+	ff_node_alloc_credit(node, cred);
 
-	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(1, node_size));
-}
+	/* credits to update two nodes : existing and newly allocated. */
+	ff_node_update_credit(node, &cred_update_node);
+	m0_be_tx_credit_mul(&cred_update_node, 2);
 
+	m0_be_tx_credit_add(cred, &cred_update_node);
+}
 
-/**
- * This function will calculate credits required for the delete KV operation and
- * add those credits to @accum.
- */
-static void btree_del_credit(const struct m0_btree  *tree, m0_bcount_t ksize,
-			     m0_bcount_t vsize, struct m0_be_tx_credit *accum)
+static void ff_put_credit(const struct nd *node, m0_bcount_t ksize,
+			  m0_bcount_t vsize, struct m0_be_tx_credit *cred)
 {
-	struct m0_be_tx_credit cred = {};
-
-	/* Credits for freeing the node. */
-	node_free_credits(tree->t_desc->t_root, ksize, vsize, &cred);
-	m0_be_tx_credit_mac(accum, &cred, MAX_TREE_HEIGHT);
+	struct m0_be_tx_credit put_cred = {};
+	/* Credits for split operation */
+	ff_node_split_credit(node, &put_cred);
```

Comment:
ff_put_credit() should only cover the credits needed for adding a record to this node. Splitting is not known at the node layer and hence node functions should not calculate credits for splitting. This should again be done at the higher layers like m0_node_put_credit().

Reply:
I think there is some confusion related to function uses. Need discussion over it.